Sources/Plasma/FeatureLib/pfConsoleCore/pfServerIni.cpp

```diff
@@ -0,0 +1,184 @@
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#include "pfServerIni.h"
+
+#include <string_theory/codecs>
+#include <string_theory/format>
+
+#include "pfConsoleParser.h"
+
+#include "pnNetBase/pnNetBase.h"
+
+#include "plFile/plEncryptedStream.h"
+
+static void ParseIntegerInto(const ST::string& value, unsigned int& out)
+{
+    ST::conversion_result res;
+    out = value.to_uint(res);
+    if (!res.ok() || !res.full_match()) {
+        throw pfServerIniParseException(ST::format("Invalid integer value: {}", value));
+    }
+}
+
+static void ParseBase64KeyInto(const ST::string& base64key, NetDhKey& output)
+{
+    ST_ssize_t base64len = ST::base64_decode(base64key, nullptr, 0);
+    if (base64len < 0) {
+        throw pfServerIniParseException("Invalid key: Invalid base-64 data (did you forget to put quotes around the value?)");
+    } else if (sizeof(output) != base64len) {
+        throw pfServerIniParseException(ST::format("Invalid key: Should be exactly {} bytes, not {}", sizeof(output), base64len));
+    }
+
+    ST_ssize_t bytes = ST::base64_decode(base64key, output, sizeof(output));
+    if (bytes < 0) {
+        throw pfServerIniParseException("Invalid key: Invalid base-64 data");
+    }
+}
+
+void pfServerIni::IParseOption(const std::vector<ST::string>& name, const ST::string& value)
+{
+    if (name.empty() || name[0].compare_i("Server") != 0) {
+        throw pfServerIniParseException("Unknown option name");
+    }
+
+    if (name.size() == 2 && name[1].compare_i("Status") == 0) {
+        fStatusUrl = value;
+    } else if (name.size() == 2 && name[1].compare_i("Signup") == 0) {
+        fSignupUrl = value;
+    } else if (name.size() == 2 && name[1].compare_i("DispName") == 0) {
+        fDisplayName = value;
+    } else if (name.size() == 2 && name[1].compare_i("Port") == 0) {
+        ParseIntegerInto(value, fPort);
+    } else if (name.size() == 3 && name[1].compare_i("File") == 0 && name[2].compare_i("Host") == 0) {
+        fFileHostname = value;
+    } else if (name.size() == 3 && name[1].compare_i("Auth") == 0) {
+        if (name[2].compare_i("Host") == 0) {
+            fAuthHostname = value;
+        } else if (name[2].compare_i("N") == 0) {
+            ParseBase64KeyInto(value, fAuthDhConstants.n);
+        } else if (name[2].compare_i("X") == 0) {
+            ParseBase64KeyInto(value, fAuthDhConstants.x);
+        } else if (name[2].compare_i("G") == 0) {
+            ParseIntegerInto(value, fAuthDhConstants.g);
+        } else {
+            throw pfServerIniParseException("Unknown option name");
```

Comment:
It might be nice to have a more specific error message to indicate which command group we encountered the error in.